src/vsphere_cpi/spec/unit/cloud/vsphere/attach_tag_to_vm_spec.rb

```diff
@@ -0,0 +1,402 @@
+           name: "fake-category-name-1"
+          )
+        )
+      end
+      let(:result_2) do
+        instance_double(VSphereAutomation::CIS::CisTaggingCategoryResult,
+          value: instance_double( VSphereAutomation::CIS::CisTaggingCategoryModel,
+            id: "fake-category-id-2",
+            name: "fake-category-name-2"
+          )
+        )
+      end
+      let(:category_ids) { %w[ fake-category-id-1 fake-category-id-2 ] }
+
+      context 'when try to find a category exists on vCenter' do
+        let(:category_name ) { "fake-category-name-1" }
+        it "should return the correct category id" do
+          allow_any_instance_of(VSphereAutomation::CIS::TaggingCategoryApi).to receive(:get).with("fake-category-id-1").and_return(result_1)
+          allow_any_instance_of(VSphereAutomation::CIS::TaggingCategoryApi).to receive(:get).with("fake-category-id-2").and_return(result_2)
+          expect(tagging_tag.retrieve_category_id(category_name, category_ids)) .to eq("fake-category-id-1")
+        end
+      end
+
+      context 'when try to find a category does not exist on vCenter' do
+        let(:category_name ) { "fake-category-name-3" }
+        it 'should return nil' do
+          allow_any_instance_of(VSphereAutomation::CIS::TaggingCategoryApi).to receive(:get).with("fake-category-id-1").and_return(result_1)
+          allow_any_instance_of(VSphereAutomation::CIS::TaggingCategoryApi).to receive(:get).with("fake-category-id-2").and_return(result_2)
+          expect(tagging_tag.retrieve_category_id(category_name, category_ids)) .to eq(nil)
+        end
+      end
+    end
+
+    describe '#retrieve_tag_id' do
+      let(:result_1) do
+        instance_double(VSphereAutomation::CIS::CisTaggingCategoryResult,
+          value: instance_double(VSphereAutomation::CIS::CisTaggingCategoryModel,
+            id: "fake-tag-id-1",
+            name: "fake-tag-name-1"
+          )
+        )
+      end
+      let(:result_2) do
+        instance_double(VSphereAutomation::CIS::CisTaggingCategoryResult,
+          value: instance_double(VSphereAutomation::CIS::CisTaggingCategoryModel,
+            id: "fake-tag-id-2",
+            name: "fake-tag-name-2"
+          )
+        )
+      end
+      let(:config_tag_1 ) { "fake-tag-name-1" }
+      let(:config_tag_2 ) { "fake-tag-name-3" }
+      let(:tag_id_list_1) { [] }
+      let(:tag_id_list_2) { %w[ fake-tag-id-1 fake-tag-id-2 ] }
+
+      context 'when received an empty tag id list' do
+        it 'should return nil' do
+          allow_any_instance_of(VSphereAutomation::CIS::TaggingCategoryApi).to receive(:get).with("fake-tag-id-1").and_return(result_1)
+          allow_any_instance_of(VSphereAutomation::CIS::TaggingCategoryApi).to receive(:get).with("fake-tag-id-2").and_return(result_2)
+          expect(tagging_tag.retrieve_category_id(config_tag_2, tag_id_list_1)).to eq(nil)
+        end
+      end
+
+      context 'when try to find a tag exists on vCenter' do
```

Comment:
`when trying to find a tag that exists on vCenter`

Reply:
Done.